morpho/processors/diagnostics/CalibrationProcessor.py

```diff
@@ -0,0 +1,278 @@
+        n_pts_below = int(round(fraction*(len(array))))
+        return 0.5*(array[n_pts_below]+array[n_pts_below-1])
+
+
+    def _get_highest_density_bounds(self, posterior_array, credibility):
+        """
+        Required input:
+            posterior_array: a list or array of posterior values
+            credibility: float between 0 and 1, denoting the credibility of the HDI.
+
+        Returns:
+            HDI: a list [p_a, p_b] containing the lower and upper value of the minimum width Bayesian credible interval
+        """
+        check_near_zero = 10
+        posterior_array.sort()
+        #Number of samples generated
+        nSample = len(posterior_array)
+        #Number of samples included in the HDI
+        nSampleCred = int(math.ceil(nSample*credibility))
+        #Number of intervals to be compared
+        nCI = nSample - nSampleCred
+        #Width of every proposed interval
+        best_width = max(posterior_array)
+        
+        for i in range(nCI):
+            if i == 0:
+                width = posterior_array[i+nSampleCred-1] - posterior_array[i]
+            else:
+                width = posterior_array[i+nSampleCred-1] - posterior_array[i]
+            if width<best_width:
+                best_width = width
+                best_index = i
+
+        #Index of lower bound of shortest interval (the HDI)
+        if (best_index == 0 and posterior_array[0]<posterior_array[check_near_zero]-posterior_array[0]):
+            HDI = [0.0, posterior_array[best_index+nSampleCred]]
+        else:
+            HDI = [posterior_array[best_index], posterior_array[best_index+nSampleCred]]
+        return HDI
+
+
+    def _report_post_param_info(self, input_vals, posterior_arrays, calib_bounds, sums):
+        """
+        For a given run in a sensitivity analysis, for some posterior credibility, determines and optionally prints the posterior median and mean.
+        Also optionally prints:
+        - Credible bounds on all parameters
+        - The parameter value inputted to the generator
+        - The posterior window (width of the C.I.)
+        """
+        for param_name in posterior_arrays:
+            b = calib_bounds[param_name]
+            median = np.median(posterior_arrays[param_name])
+            sums[param_name]['median'] += median
+            mean = np.mean(posterior_arrays[param_name])
+            sums[param_name]['mean'] += mean
+
+            if len(self.cred_interval) == 1:
+                if self.verbose==True:
+                    logger.info("{} < {}. Input: {}".format(param_name, b[0], input_vals[param_name]))
+            elif len(self.cred_interval) == 2:
+                sums[param_name]['lower'] += b[0]
+                sums[param_name]['upper'] += b[1]
+                if self.verbose==True:
+                    logger.info("{} < {} < {}, Median={}, Mean={}. Input value: {}. Window: {}".format(b[0], param_name, b[1], median, mean, input_vals[param_name], b[1]-b[0]))
+            else:
```

Comment:
Wouldn't it make more sense to print this "please input..." during the Configure part?
It is more likely to be visible there than at the very end of the Run part...

Reply:
You're right – can/should I edit the version of this processor that's now in `develop`?

Reply:
Yes, I was not sure what to do.
Please do and we will make a new release.

Reply:
I think I should tell the processor not to run if InternalConfigure determines that cred_interval is the wrong length. What is the proper way to do that?

Reply:
Right now, you can check the length and return False if this check fails: this works for every kind of tests.
Then in your script, you shoud check the returned value from processor.Configure(): if false, then the script should stop.
Does it make sense?

Feel free to add this in the feature/calib_fixes branch, so it gets included there!

Later on (with a proper Exception system), this will be taken care using `raise Exception`...

Reply:
I made the change here: https://github.com/morphoorg/morpho/pull/149/commits/603ad0c09345a545600294048bdf1d11a1193494